draft-ietf-mls-protocol.md

```diff
@@ -1435,48 +1461,210 @@ all arrive at the following state:
 
 # Message Protection
 
-[[ OPEN ISSUE: This section has initial considerations about message
-protection.  This issue clearly needs more specific recommendations,
-possibly a protocol specification in this document or a separate
-one. ]]
+The primary purpose of the handshake protocol is to provide an authenticated
+group key exchange to participants. In order to protect Application messages
+sent among those participants, the Application secret provided by the Handshake
+key schedule is used to derive encryption keys for the Message Protection Layer.
+
+Application messages MUST be protected with the Authenticated-Encryption
+with Associated-Data (AEAD) encryption scheme associated with the MLS ciphersuite.
+Note that "Authenticated" in this context does not mean messages are known to
+be sent by a specific participant but only from a legitimate member of the group.
+To authenticate a message from a particular member, signatures are required.
+Handshake messages MUST use asymmetric signatures to strongly authenticate
+the sender of a message.
+
+Each participant maintains their own chain of Application secrets, where the first
+one is derived based on a secret chained from the Epoch secret.
+As shown in {{key-schedule}}, the initial Application secret is bound to the
+identity of each participant to avoid collisions and allow support for decryption
+of reordered messages.
+
+Subsequent Application secrets MUST be rotated for each message sent in
+order to provide stronger cryptographic security guarantees. The Application
+Key Schedule use this rotation to generate fresh AEAD encryption keys and nonces
+used to encrypt and decrypt future Application messages.
+In all cases, a participant MUST NOT encrypt more than expected by the security
+bounds of the AEAD scheme used.
+
+Note that each change to the Group through a Handshake message will cause
+a change of the Group Secret. Hence this change MUST be applied before encrypting
+any new Application message. This is required for confidentiality reasons
+in order for Members to avoid receiving messages from the group after leaving,
+being added to, or excluded from the Group.
+
+## Application Key Schedule {#key-schedule-application}
+
+Updating the Application secret and deriving the associated AEAD key and nonce can
+be summarized as the following Application key schedule where
+each participant's Application secret chain looks as follows after the initial
+derivation:
 
-The primary purpose of this protocol is to enable an authenticated
-group key exchange among participants.  In order to protect messages sent among
-those participants, an application will need to specify how messages are protected.
+~~~~~
+           application_secret_N-1
+                     |
+                     +--> HKDF-Expand-Label(.,"nonce", "", nonce_length)
```

Comment:
Actually, it occurs to me that you have a simultaneous-send issue here -- two senders can chose to transmit based on the same application secret, and then you have nonce reuse, which I hear is bad.

At the IETF meeting, we had discussed two potential approaches here:

1. Per-sender hash ratchets
2. A single hash ratchet for the group folding in sender information at key/nonce generation time

You need to do one of those here.  I continue to prefer the second one, and it is a smaller delta on what you have here.  You would just need to change the "" in the nonce/key derivations to some encoding of the sender ID.

Reply:
No they can't do it, maybe it is not clear enough but this is the per-sender chaining here, so the Participant application secret is used...

Reply:
"each participant's Application secret chain looks as follows after the initial
derivation:"